In [1]:
from IPython.display import display, HTML

display(HTML("""
<style>
    /* 1. UMUMIY KONTEYNER VA SHRIFT */
    .rendered_html, .jp-RenderedHTMLCommon, .jp-RenderedMarkdown {
        font-family: "Times New Roman", Times, serif !important;
        color: black !important;
        font-feature-settings: "kern" 1, "liga" 1, "clig" 1, "dlig" 1 !important;
        -webkit-font-smoothing: antialiased !important;
    }

    /* 2. MATNLAR: 1.5 oraliq, HAR BIR abzats 1.25 sm chekinadi */
    .rendered_html p, .jp-RenderedHTMLCommon p {
        font-family: "Times New Roman", Times, serif !important;
        font-size: 14pt !important;
        line-height: 1.5 !important;
        text-align: justify !important;
        text-indent: 1.25cm !important;
        margin: 0 !important;
        hyphens: auto !important;
    }

    /* Ro'yxatlar (nuqtalar ko'rinishi) */
    .rendered_html ul, .jp-RenderedHTMLCommon ul {
        list-style-type: disc !important;
        margin-left: 1.5cm !important;
        padding: 0 !important;
        text-indent: 0 !important;
    }
    .rendered_html ol, .jp-RenderedHTMLCommon ol {
        margin-left: 1.5cm !important;
        padding: 0 !important;
        text-indent: 0 !important;
    }
    .rendered_html li, .jp-RenderedHTMLCommon li {
        display: list-item !important;
        font-family: "Times New Roman", Times, serif !important;
        font-size: 14pt !important;
        line-height: 1.5 !important;
        text-indent: 0 !important;
        text-align: justify !important;
        margin: 0 !important;
    }

    /* 3. SARLAVHALAR: h1 markazda; h2-h4 chapda, 1.25 sm chekinadi */
    .rendered_html h1, .jp-RenderedHTMLCommon h1,
    .rendered_html h2, .jp-RenderedHTMLCommon h2,
    .rendered_html h3, .jp-RenderedHTMLCommon h3,
    .rendered_html h4, .jp-RenderedHTMLCommon h4 {
        font-family: "Times New Roman", Times, serif !important;
        margin: 0.6em 0 0.3em 0 !important;
        font-weight: bold !important;
    }
    .rendered_html h1, .jp-RenderedHTMLCommon h1 {
        font-size: 20pt !important; text-align: center !important; text-indent: 0 !important;
    }
    .rendered_html h2, .jp-RenderedHTMLCommon h2 {
        font-size: 16pt !important; text-align: left !important; text-indent: 1.25cm !important;
    }
    .rendered_html h3, .jp-RenderedHTMLCommon h3 {
        font-size: 14pt !important; text-align: left !important; text-indent: 1.25cm !important;
    }
    .rendered_html h4, .jp-RenderedHTMLCommon h4 {
        font-size: 13pt !important; text-align: left !important; text-indent: 1.25cm !important;
    }

    /* 4. "In[ ]" / "Out[ ]" PROMPTLARINI YASHIRISH */
    .jp-InputPrompt, .jp-OutputPrompt, .prompt, .input_prompt, .output_prompt {
        display: none !important;
        width: 0 !important;
        visibility: hidden !important;
    }

    /* Kod bloklarini chap chekkaga taqash */
    .jp-CodeCell .jp-Cell-inputWrapper, .cell.code_cell .input {
        margin-left: 0 !important;
        padding-left: 0 !important;
    }

    /* Kod shrifti */
    .jp-CodeMirror, .CodeMirror, .CodeMirror-scroll, .CodeMirror-sizer, .CodeMirror-lines,
    .jp-OutputArea-output pre, .output_area pre,
    .CodeMirror-code, .CodeMirror-line,
    code, kbd, samp {
        font-family: "Consolas", monospace !important;
        font-size: 12pt !important;
        text-indent: 0 !important;
        line-height: 1.3 !important;
    }

    /* BLOK kod: chiroyli o'ralsin (sahifadan chiqmasin) */
    .jp-OutputArea-output pre, .output_area pre,
    .CodeMirror-line {
        white-space: pre-wrap !important;
        overflow-wrap: anywhere !important;
        word-break: normal !important;
        overflow-x: visible !important;
        height: auto !important;
    }

    /* MATN ICHIDAGI kod: bir qatorda qolsin */
    :not(pre) > code, kbd, samp {
        white-space: nowrap !important;
    }

    /* 5. BIRINCHI KATAKNI (shu stil kodini) YASHIRISH, tanlanganda ochish */
    .jp-Notebook .jp-Cell:first-child:not(.jp-mod-selected),
    .container div.cell.code_cell:first-of-type:not(.selected) {
        position: absolute !important;
        height: 0 !important;
        min-height: 0 !important;
        width: 0 !important;
        margin: 0 !important;
        padding: 0 !important;
        border: none !important;
        overflow: hidden !important;
        opacity: 0 !important;
        pointer-events: none !important;
        visibility: hidden !important;
    }
    .jp-Notebook .jp-Cell:first-child.jp-mod-selected,
    .container div.cell.code_cell:first-of-type.selected {
        position: relative !important;
        height: auto !important;
        width: 100% !important;
        opacity: 1 !important;
        overflow: visible !important;
        pointer-events: auto !important;
        visibility: visible !important;
    }

    /* JADVALLAR */
    .rendered_html table, .jp-RenderedHTMLCommon table,
    .rendered_html td, .jp-RenderedHTMLCommon td,
    .rendered_html th, .jp-RenderedHTMLCommon th {
        font-family: "Times New Roman", Times, serif !important;
        font-size: 14pt !important;
        line-height: 1.5 !important;
        color: black !important;
    }
</style>
"""))

# 2. Algebraik strukturalar

Algebra dastlab tenglamalarni yechish san’ati sifatida paydo bo‘lgan. Vaqt o‘tishi bilan matematiklar bir narsani payqashdi: turli sohalardagi hisob-kitoblar — butun sonlarni qo‘shish, matritsalarni ko‘paytirish, geometrik shakllarni burish — bir xil qonuniyatlarga bo‘ysunadi. Shu umumiy qonuniyatlarni ajratib olib, alohida o‘rganish g‘oyasi **algebraik struktura** tushunchasini tug‘dirdi.

Algebraik struktura deganda biror **to‘plam** va shu to‘plam elementlari ustida bajariladigan bir yoki bir necha **amal**, hamda shu amallar bo‘ysunadigan **qoidalar (aksiomalar)** majmuasi tushuniladi. Bu bobda biz eng muhim algebraik strukturalarni — yarim gruppa, monoid, gruppa, halqa va maydonni — sodda misollar orqali, bosqichma-bosqich ko‘rib chiqamiz.

Bu tushunchalar zamonaviy kriptografiyaning tayanchidir. Shifrlash algoritmlari ma’lumotni odatda **chekli to‘plam** elementlari (masalan, baytlar yoki sonlar) sifatida qayta ishlaydi. Shifrlash va deshifrlash amallari natijasi doimo yana o‘sha to‘plamga tegishli bo‘lishi — ya’ni to‘plam amalga nisbatan **yopiq** bo‘lishi — shart. Aynan shu talab bizni gruppa, halqa va maydon kabi strukturalarga olib keladi.<sup>8</sup>

## 2.1 Binar amal

Har qanday algebraik strukturaning poydevori — **amal** tushunchasi. Ko‘pincha amal ikki elementni oladi va ulardan bittasini hosil qiladi (masalan, $2 + 3 = 5$ da qo‘shish $2$ va $3$ dan $5$ ni hosil qiladi). Bunday amal **binar amal** (ikki o‘rinli amal) deyiladi.

**Ta’rif.** $S$ — to‘plam bo‘lsin. $S$ da aniqlangan **binar amal** deb, $S$ ning har bir tartiblangan $(a, b)$ juftligiga $S$ ning aniq bir elementini mos qo‘yuvchi qoidaga aytiladi.<sup>1</sup> Binar amalni odatda $*$, $\circ$, $\oplus$, $\cdot$ kabi belgilar bilan yozamiz; $(a, b)$ juftlikka mos kelgan element $a * b$ ko‘rinishida belgilanadi.

Bu ta’rifda muhim bir yashirin shart bor: $a * b$ natijasi **yana $S$ ning elementi** bo‘lishi kerak. Bu xossa **yopiqlik** deyiladi.<sup>1</sup> Masalan, natural sonlar to‘plami $\mathbb{N} = \{1, 2, 3, \dots\}$ qo‘shishga nisbatan yopiq (ikki natural sonning yig‘indisi yana natural son), lekin ayirishga nisbatan yopiq emas, chunki $2 - 5 = -3 \notin \mathbb{N}$.

### Amalning xossalari

Binar amallar ba’zi qo‘shimcha xossalarga ega bo‘lishi mumkin. Quyidagi ikki xossa keyingi barcha ta’riflarda uchraydi.

**Assotsiativlik (guruhlash xossasi).** $*$ amali $S$ da assotsiativ deyiladi, agar barcha $a, b, c \in S$ uchun

$$(a * b) * c = a * (b * c)$$

tenglik o‘rinli bo‘lsa.<sup>1</sup> Ma’nosi: uch (yoki undan ko‘p) elementni birlashtirganda, qavslarni qanday joylashtirishimiz natijaga ta’sir qilmaydi. Shu sababli assotsiativ amalda qavslarni umuman tushirib qoldirib, $a * b * c$ deb yozaverish mumkin. Masalan, $(2 + 3) + 4 = 2 + (3 + 4) = 9$.

**Kommutativlik (o‘rin almashtirish xossasi).** $*$ amali $S$ da kommutativ deyiladi, agar barcha $a, b \in S$ uchun

$$a * b = b * a$$

bo‘lsa.<sup>1</sup> Ma’nosi: elementlarning tartibi natijaga ta’sir qilmaydi. Qo‘shish va ko‘paytirish kommutativ ($3 + 5 = 5 + 3$), ammo, masalan, matritsalarni ko‘paytirish kommutativ emas.

> **Diqqat.** Assotsiativlik va kommutativlik — bir-biridan mustaqil ikki xossa. Matritsalarni ko‘paytirish assotsiativ, lekin kommutativ emas.<sup>8</sup> Aksincha bo‘lgan (kommutativ, ammo assotsiativ bo‘lmagan) amallar ham mavjud.

### Amallar jadvali (Keli jadvali)

To‘plam **chekli** bo‘lsa, undagi amalni jadval ko‘rinishida to‘liq berish qulay. Bunday jadval **amallar jadvali** yoki **Keli jadvali** deyiladi.<sup>2</sup> Jadvalning $a$-satri va $b$-ustuni kesishmasiga $a * b$ natijasi yoziladi.

Masalan, $\mathbb{Z}_4 = \{0, 1, 2, 3\}$ to‘plamida **4 ga bo‘lgandagi qoldiq bo‘yicha qo‘shish** amali (uni $\oplus$ deb belgilaymiz, masalan $2 \oplus 3 = 5 \bmod 4 = 1$) quyidagicha:

| $\oplus$ | 0 | 1 | 2 | 3 |
|:---:|:---:|:---:|:---:|:---:|
| **0** | 0 | 1 | 2 | 3 |
| **1** | 1 | 2 | 3 | 0 |
| **2** | 2 | 3 | 0 | 1 |
| **3** | 3 | 0 | 1 | 2 |

Jadvaldan xossalarni ko‘z bilan o‘qish mumkin: jadval asosiy diagonalga nisbatan simmetrik bo‘lsa — amal kommutativ; har bir element har bir satr va ustunda aynan bir marta uchrasa — bu strukturaning muhim belgisi (buni gruppa bo‘limida ko‘ramiz).

## 2.2 Yarim gruppa va monoidlar

Endi to‘plamga amalni bog‘lab, birinchi algebraik strukturalarni quramiz. Ularni murakkabligi oshib boruvchi zinapoya sifatida tasavvur qiling: har bir keyingi struktura oldingisiga bitta yangi shart qo‘shadi.

### Yarim gruppa

**Ta’rif.** $(S, *)$ juftlik **yarim gruppa** deyiladi, agar:

1. $S$ — bo‘sh bo‘lmagan to‘plam va $*$ — $S$ da aniqlangan binar amal (ya’ni $S$ amalga nisbatan **yopiq**);
2. $*$ amali **assotsiativ**, ya’ni barcha $a, b, c \in S$ uchun $(a * b) * c = a * (b * c)$.<sup>1, 8</sup>

Ya’ni yarim gruppa — bu shunchaki **assotsiativ binar amalga ega to‘plam**. Boshqa hech qanday shart talab qilinmaydi.

**Misol.** Natural sonlar to‘plami $\mathbb{N} = \{1, 2, 3, \dots\}$ qo‘shishga nisbatan yarim gruppa: ikki natural sonning yig‘indisi yana natural son (yopiqlik) va qo‘shish assotsiativ. Xuddi shunday, $\mathbb{N}$ ko‘paytirishga nisbatan ham yarim gruppa.

**Misol.** Biror alifbodagi bo‘sh bo‘lmagan **so‘zlar (satrlar)** to‘plami, amal sifatida **birlashtirish (konkatenatsiya)** olinsa, yarim gruppa hosil bo‘ladi: `"ax"` va `"by"` satrlarni birlashtirsak `"axby"` hosil bo‘ladi, va birlashtirish assotsiativdir.<sup>3</sup> Bu misol xesh-funksiyalar ishlashini tushunishda muhim, chunki uzun xabar qisqa bloklarning birlashmasi sifatida qaraladi.

### Monoid

Yarim gruppaga bitta tabiiy shart — **neytral element**ning mavjudligi — qo‘shsak, monoid hosil bo‘ladi.

**Ta’rif.** $(S, *)$ yarim gruppa **monoid** deyiladi, agar $S$ da shunday **neytral element** $e$ mavjud bo‘lsaki, barcha $a \in S$ uchun

$$a * e = e * a = a$$

o‘rinli bo‘lsin.<sup>3, 8</sup> Neytral element — amalni bajarganda elementni “o‘zgartirmaydigan” element.

**Misol.** $\mathbb{N}_0 = \{0, 1, 2, \dots\}$ (nol qo‘shilgan natural sonlar) qo‘shishga nisbatan monoid: neytral element $0$, chunki $a + 0 = a$. Ko‘paytirishga nisbatan esa neytral element $1$, chunki $a \cdot 1 = a$; demak $(\mathbb{N}, \cdot)$ ham monoid.

**Misol.** Satrlar to‘plamiga **bo‘sh satr** `""` ni qo‘shsak, birlashtirishga nisbatan monoid hosil bo‘ladi: bo‘sh satrni istalgan satrga ulasak, satr o‘zgarmaydi (`"" + "ab" = "ab"`), demak bo‘sh satr — neytral element.<sup>3</sup>

> **Muhim izoh — neytral element yagona.** Agar monoidda $e$ va $e'$ ikkita neytral element bo‘lsa, ularni bir-biriga qo‘llab: $e = e * e' = e'$ (birinchi tenglikda $e'$ neytral, ikkinchisida $e$ neytral bo‘lgani ishlatildi). Demak $e = e'$ — neytral element **hamisha yagona**.<sup>3</sup>

### Nega bu axborot xavfsizligida muhim?

Ko‘p kriptografik konstruksiyalar aynan yarim gruppa yoki monoid tuzilishiga tayanadi.

Xesh-funksiyalar (masalan, SHA-256) uzun xabarni qat’iy uzunlikdagi **bloklarga** ajratib, bloklarni ketma-ket qayta ishlaydi. Xabar — bloklarning birlashmasi; birlashtirish esa assotsiativ amal. Masalan, $1000$ baytli xabarni $64$ baytli $16$ ta blokka bo‘lish, so‘ng ularni ma’lum tartibda qayta ishlash — bu monoiddagi ketma-ket amaldan boshqa narsa emas: bloklarni guruhlash tartibi (assotsiativlik tufayli) natijaga ta’sir qilmaydi, bo‘sh xabar esa neytral element rolini o‘ynaydi.

Diqqat qilaylik: satrlarni birlashtirish monoidida **teskari element yo‘q** — `"ab"` ni birlashtirish orqali `""` ga qaytarib bo‘lmaydi.<sup>3</sup> Aynan shu “orqaga qaytib bo‘lmaslik” xususiyati bir tomonlama funksiyalar mohiyatiga yaqin turadi. Teskari elementga ega bo‘lish uchun bizga kuchliroq struktura — **gruppa** kerak bo‘ladi.


In [2]:
# Chekli to‘plamda binar amalni tekshirish: (Z_4, mod 4 bo‘yicha qo‘shish)
def cayley(n, op):
    print("   " + " ".join(f"{j:2d}" for j in range(n)))
    for i in range(n):
        print(f"{i:2d}|" + " ".join(f"{op(i, j):2d}" for j in range(n)))

n = 4
op = lambda a, b: (a + b) % n
print("Z_4 ustida ⊕ (mod 4 qo‘shish) amallar jadvali:")
cayley(n, op)

# Xossalarni dastur bilan tekshiramiz
S = range(n)
yopiq  = all(op(a, b) in S for a in S for b in S)
assots = all(op(op(a, b), c) == op(a, op(b, c)) for a in S for b in S for c in S)
kommut = all(op(a, b) == op(b, a) for a in S for b in S)
print(f"\nYopiqlik: {yopiq} | Assotsiativlik: {assots} | Kommutativlik: {kommut}")

Z_4 ustida ⊕ (mod 4 qo‘shish) amallar jadvali:
    0  1  2  3
 0| 0  1  2  3
 1| 1  2  3  0
 2| 2  3  0  1
 3| 3  0  1  2

Yopiqlik: True | Assotsiativlik: True | Kommutativlik: True


## 2.3 Gruppa: ta’rif va misollar

Monoidga yana bitta shart — **har bir element uchun teskari elementning mavjudligi** — qo‘shsak, algebraning eng markaziy tushunchasi bo‘lmish **gruppa** hosil bo‘ladi. Gruppa — bu shunday struktura bo‘lib, unda amalni nafaqat bajarish, balki uni **“bekor qilish” (orqaga qaytarish)** ham mumkin.

**Ta’rif.** $G$ to‘plam va undagi $*$ binar amal **gruppa** tashkil etadi, agar quyidagi to‘rt shart bajarilsa:<sup>2</sup>

1. **Yopiqlik.** Barcha $a, b \in G$ uchun $a * b \in G$.
2. **Assotsiativlik.** Barcha $a, b, c \in G$ uchun $(a * b) * c = a * (b * c)$.
3. **Neytral element.** $G$ da shunday $e$ element mavjudki, barcha $a \in G$ uchun $e * a = a * e = a$.
4. **Teskari element.** Har bir $a \in G$ uchun $G$ da shunday $a^{-1}$ element mavjudki, $a * a^{-1} = a^{-1} * a = e$.

Ushbu ta’rifni zinapoya sifatida ko‘rish foydali:

- faqat 1- va 2-shart bajarilsa — **yarim gruppa**;
- 1-, 2-, 3-shartlar bajarilsa — **monoid**;
- to‘rttala shart bajarilsa — **gruppa**.<sup>8</sup>

**Ta’rif.** Agar gruppada, qo‘shimcha ravishda, amal **kommutativ** ($a * b = b * a$) bo‘lsa, u **Abel gruppasi** yoki **kommutativ gruppa** deyiladi.<sup>2</sup> Nom norvegiyalik matematik Nils Henrik Abel sharafiga qo‘yilgan.

### Gruppaning oddiy xossalari

Ta’rifning o‘zidan bir necha muhim natija kelib chiqadi. Neytral element yagona bo‘lgani kabi (buni monoidda ko‘rgan edik), **har bir elementning teskarisi ham yagona**.<sup>3</sup> Haqiqatan, $a$ ning $a'$ va $a''$ ikkita teskarisi bo‘lsa:

$$a' = a' * e = a' * (a * a'') = (a' * a) * a'' = e * a'' = a''.$$

Bu yerda ketma-ket neytral element ta’rifi, teskari element ta’rifi va assotsiativlik ishlatildi. Demak $a' = a''$.

### Misollar

**Misol 1 — Butun sonlar $(\mathbb{Z}, +)$.** Butun sonlar to‘plami qo‘shishga nisbatan Abel gruppasi:<sup>2</sup>

- yopiqlik: ikki butun sonning yig‘indisi — butun son;
- assotsiativlik: $(a + b) + c = a + (b + c)$;
- neytral element: $0$, chunki $a + 0 = a$;
- teskari element: $a$ ning teskarisi $-a$, chunki $a + (-a) = 0$;
- kommutativlik: $a + b = b + a$.

Diqqat: $(\mathbb{Z}, \cdot)$, ya’ni butun sonlar **ko‘paytirishga** nisbatan gruppa **emas** — $0$ ning teskarisi yo‘q, hatto $2$ ning ham teskarisi ($1/2$) butun son emas.

**Misol 2 — Qoldiqlar gruppasi $(\mathbb{Z}_n, +)$.** $\mathbb{Z}_n = \{0, 1, \dots, n-1\}$ to‘plami **$n$ ga bo‘lgandagi qoldiq bo‘yicha qo‘shish**ga nisbatan Abel gruppasi.<sup>2</sup> Bu yerda neytral element $0$, $a$ ning teskarisi esa $n - a$ (chunki $a + (n-a) = n \equiv 0 \pmod{n}$). Masalan $\mathbb{Z}_5$ da $2$ ning teskarisi $3$, chunki $2 + 3 = 5 \equiv 0$.

**Misol 3 — Teskari qoldiqlar gruppasi $U(n)$.** $\mathbb{Z}_n$ ni **ko‘paytirishga** nisbatan qaralsa, hamma element teskariga ega bo‘lavermaydi. $n$ bilan **o‘zaro tub** ($\gcd(a, n) = 1$) bo‘lgan $a$ largina teskariga ega. Ana shunday elementlar to‘plami $U(n)$ ko‘paytirishga nisbatan Abel gruppasi tashkil etadi.<sup>2</sup> Masalan, $U(8) = \{1, 3, 5, 7\}$, chunki bu sonlarning har biri $8$ bilan o‘zaro tub va $3 \cdot 3 = 9 \equiv 1$, $5 \cdot 5 = 25 \equiv 1$, $7 \cdot 7 = 49 \equiv 1 \pmod 8$ — har biri o‘z-o‘ziga teskari.

**Misol 4 — Noldan farqli kompleks sonlar $(\mathbb{C}^{*}, \cdot)$.** Noldan farqli kompleks sonlar ko‘paytirishga nisbatan Abel gruppasi; masalan $H = \{1, -1, i, -i\}$ ham shu gruppaning kichik gruppasi bo‘lib, ular aynan $z^4 = 1$ tenglamaning yechimlaridir.<sup>2</sup> Bu bilan 2.7-bo‘limda yana uchrashamiz.

**Misol 5 — Kommutativ bo‘lmagan gruppa.** Determinanti $\pm 1$ bo‘lgan $2 \times 2$ butun sonli matritsalar, ko‘paytirishga nisbatan, gruppa hosil qiladi — ammo matritsalarni ko‘paytirish o‘rin almashmaydigani uchun bu **Abel bo‘lmagan** gruppadir.<sup>3</sup> Uch va undan ko‘p elementli to‘plamning **o‘rniga qo‘yishlari (perestanovkalari)**, kompozitsiya amaliga nisbatan, yana Abel bo‘lmagan gruppaga misol.<sup>3</sup>

### Nega bu axborot xavfsizligida muhim?

Zamonaviy ochiq kalitli kriptografiyaning katta qismi aynan gruppalar, xususan $\mathbb{Z}_p^{*}$ (tub $p$ modul bo‘yicha teskari qoldiqlar gruppasi) ustida quriladi. Teskari elementning mavjudligi deshifrlash imkonini beradi, gruppadagi **darajaga ko‘tarish** amali esa **Diffie–Hellman** kalit almashinuvi va **ElGamal** shifrlashning asosidir.

Aniq sonli misol ko‘raylik. $p = 23$ tub sonni va $g = 5$ generatorni ochiqdan tanlaymiz; bularning ustida $\mathbb{Z}_{23}^{*}$ gruppasida ishlaymiz.

- Alisa maxfiy $a = 6$ ni tanlaydi va $A = g^{a} \bmod 23 = 5^{6} \bmod 23 = 8$ ni yuboradi.
- Bob maxfiy $b = 15$ ni tanlaydi va $B = g^{b} \bmod 23 = 5^{15} \bmod 23 = 19$ ni yuboradi.
- Alisa umumiy kalitni $B^{a} \bmod 23 = 19^{6} \bmod 23 = 2$ deb hisoblaydi.
- Bob esa $A^{b} \bmod 23 = 8^{15} \bmod 23 = 2$ ni oladi.

Ikkalasi ham bir xil $2$ maxfiy kalitiga keldi, garchi hech qachon maxfiy sonlarini ochiq uzatmagan bo‘lsalar ham. Buzg‘unchi $g$, $p$, $A$, $B$ ni bilsa-da, umumiy kalitni topish uchun $g^{x} \equiv A \pmod p$ tenglamadagi $x$ ni topishi — **diskret logarifm** masalasini yechishi kerak bo‘ladi; katta $p$ larda esa bu amaliy jihatdan bajarib bo‘lmas darajada qiyin. Butun tizimning ishonchliligi shu gruppaviy xossalarga tayanadi.


In [3]:
from math import gcd

def guruh_aksiomalari(elems, op, e):
    yopiq  = all(op(a, b) in elems for a in elems for b in elems)
    neytr  = all(op(e, a) == a and op(a, e) == a for a in elems)
    teskr  = all(any(op(a, b) == e and op(b, a) == e for b in elems) for a in elems)
    kommut = all(op(a, b) == op(b, a) for a in elems for b in elems)
    return yopiq, neytr, teskr, kommut

# U(8) — 8 bilan o‘zaro tub qoldiqlar, mod 8 ko‘paytirish bo‘yicha gruppa
n = 8
U = [a for a in range(1, n) if gcd(a, n) == 1]
print("U(8) =", U)
y, nt, t, k = guruh_aksiomalari(U, lambda a, b: (a * b) % n, 1)
print(f"Yopiq={y}, neytral(1)={nt}, teskarili={t}, kommutativ={k}")
for a in U:
    inv = next(b for b in U if (a * b) % n == 1)
    print(f"  {a}⁻¹ = {inv}  (chunki {a}·{inv} = {a*inv} ≡ 1 mod 8)")

print("\n--- Diffie–Hellman (Z_23*, g=5) ---")
p, g = 23, 5
a, b = 6, 15                       # maxfiy sonlar
A, B = pow(g, a, p), pow(g, b, p)  # ochiq uzatiladi
print(f"Alisa yuboradi A = 5^{a} mod 23 = {A}")
print(f"Bob   yuboradi B = 5^{b} mod 23 = {B}")
kalit_alisa = pow(B, a, p)
kalit_bob   = pow(A, b, p)
print(f"Alisaning kaliti = B^{a} mod 23 = {kalit_alisa}")
print(f"Bobning  kaliti = A^{b} mod 23 = {kalit_bob}")
print("Umumiy maxfiy kalit mos keldi:", kalit_alisa == kalit_bob)

U(8) = [1, 3, 5, 7]
Yopiq=True, neytral(1)=True, teskarili=True, kommutativ=True
  1⁻¹ = 1  (chunki 1·1 = 1 ≡ 1 mod 8)
  3⁻¹ = 3  (chunki 3·3 = 9 ≡ 1 mod 8)
  5⁻¹ = 5  (chunki 5·5 = 25 ≡ 1 mod 8)
  7⁻¹ = 7  (chunki 7·7 = 49 ≡ 1 mod 8)

--- Diffie–Hellman (Z_23*, g=5) ---
Alisa yuboradi A = 5^6 mod 23 = 8
Bob   yuboradi B = 5^15 mod 23 = 19
Alisaning kaliti = B^6 mod 23 = 2
Bobning  kaliti = A^15 mod 23 = 2
Umumiy maxfiy kalit mos keldi: True


## 2.4 Halqa va maydonlar: misollar

Hozirgacha biz **bitta** amalga ega strukturalarni ko‘rdik. Ammo tabiiy sanoq sohalarida — masalan butun sonlarda — bir vaqtning o‘zida **ikkita** amal, qo‘shish va ko‘paytirish, mavjud bo‘lib, ular bir-biri bilan **distributivlik qonuni** orqali bog‘langan. Ikki bunday bog‘langan amalga ega strukturalar **halqa** va **maydon** deyiladi.<sup>4</sup>

### Halqa

**Ta’rif.** Bo‘sh bo‘lmagan $R$ to‘plam **halqa** deyiladi, agar unda ikkita yopiq binar amal — **qo‘shish** ($+$) va **ko‘paytirish** ($\cdot$) — aniqlangan bo‘lib, quyidagi shartlar bajarilsa:<sup>4</sup>

1. $R$ qo‘shishga nisbatan **Abel gruppasi** (ya’ni qo‘shish assotsiativ, kommutativ, neytral element $0$ mavjud, har bir $a$ uchun qarama-qarshi element $-a$ mavjud);
2. ko‘paytirish **assotsiativ**: $(a \cdot b) \cdot c = a \cdot (b \cdot c)$;
3. ko‘paytirish qo‘shishga nisbatan **distributiv**:

$$a \cdot (b + c) = a \cdot b + a \cdot c, \qquad (a + b) \cdot c = a \cdot c + b \cdot c.$$

Distributivlik — bu ikki amalni bir-biriga bog‘lovchi asosiy qonun; usiz halqa oddiygina “ikkita alohida struktura” bo‘lib qolardi.

Halqada ba’zi qo‘shimcha xossalar uchrashi mumkin, va ular halqalarni turlarga ajratadi:<sup>4</sup>

- Agar $R$ da ko‘paytirish uchun ham **birlik element** $1$ mavjud bo‘lsa ($1 \ne 0$ va $1 \cdot a = a \cdot 1 = a$), $R$ — **birlikli halqa** deyiladi.
- Agar ko‘paytirish **kommutativ** ($a \cdot b = b \cdot a$) bo‘lsa, $R$ — **kommutativ halqa**.
- Kommutativ birlikli halqada, agar $a \cdot b = 0$ dan doimo $a = 0$ yoki $b = 0$ kelib chiqsa (ya’ni **nolning bo‘luvchilari yo‘q** bo‘lsa), $R$ — **butunlik sohasi** deyiladi.

**Misol.** Butun sonlar $(\mathbb{Z}, +, \cdot)$ — kommutativ birlikli halqa, hatto butunlik sohasi: ikki noldan farqli butun sonning ko‘paytmasi noldan farqli.<sup>4</sup> Ma’ruza matnida ko‘rsatilganidek, $\mathbb{Z}$ da $+$ Abel gruppasi, $\cdot$ assotsiativ va distributiv — demak $(\mathbb{Z}, +, \cdot)$ struktura halqadir.<sup>8</sup>

### Maydon

Maydon — bu halqalar ichida eng “to‘kis”i: unda nafaqat qo‘shish, balki (noldan tashqari) **ko‘paytirish ham to‘liq teskarilanadi**.

**Ta’rif.** Kommutativ birlikli halqa $F$ **maydon** deyiladi, agar uning har bir **noldan farqli** elementi ko‘paytirishga nisbatan teskariga ega bo‘lsa. Ya’ni har bir $a \ne 0$ uchun shunday $a^{-1} \in F$ mavjudki, $a \cdot a^{-1} = 1$.<sup>4</sup>

Boshqacha aytganda, maydonda **ikkala** amal ham (nol elementi bundan mustasno) gruppa tuzadi: $F$ qo‘shishga nisbatan Abel gruppasi, $F \setminus \{0\}$ esa ko‘paytirishga nisbatan Abel gruppasi. Aynan shu ikki tomonlama to‘liqlik maydonlarni chiziqli tenglamalarni yechish uchun ideal muhitga aylantiradi.

Ma’ruzada keltirilgan maydonning muhim xossalari:<sup>8</sup> (1) har bir maydonda birlik element yagona; (2) har bir $x \ne 0$ uchun teskari $x^{-1}$ yagona; (3) maydonda nolning bo‘luvchilari yo‘q — agar $x \cdot y = 0$ bo‘lsa, $x = 0$ yoki $y = 0$.

**Misol — sonli maydonlar.** Ratsional sonlar $\mathbb{Q}$, haqiqiy sonlar $\mathbb{R}$ va kompleks sonlar $\mathbb{C}$ — maydonlar.<sup>4</sup> Butun sonlar $\mathbb{Z}$ esa maydon **emas**, chunki, masalan, $2$ ning ko‘paytmaviy teskarisi $\tfrac12$ butun son emas.

**Misol — chekli maydon $\mathbb{Z}_p$.** Agar $p$ — tub son bo‘lsa, $\mathbb{Z}_p = \{0, 1, \dots, p-1\}$ mod $p$ amallariga nisbatan maydon bo‘ladi.<sup>4</sup> Masalan $\mathbb{Z}_5$ da $2$ ning teskarisi $3$, chunki $2 \cdot 3 = 6 \equiv 1 \pmod 5$. Ammo $p$ tub bo‘lmasa, bu buziladi: $\mathbb{Z}_6$ da $2 \cdot 3 = 6 \equiv 0$ — noldan farqli ikki elementning ko‘paytmasi nol, demak $\mathbb{Z}_6$ hatto butunlik sohasi ham emas.

**Misol — $2 \times 2$ maydon.** $\mathbb{Z}_2$ ustidagi ba’zi $2 \times 2$ matritsalar to‘plami ham chekli maydon tuzishi mumkin — bu chekli maydonlar faqat $\mathbb{Z}_p$ ko‘rinishida bo‘lmasligini ko‘rsatadi.<sup>4</sup>

### Nega bu axborot xavfsizligida muhim?

Halqa va maydonlar zamonaviy simmetrik va asimmetrik shifrlashning bevosita matematik asosidir.

**RSA algoritmi** $\mathbb{Z}_n$ **halqasida** ishlaydi, bunda $n = p \cdot q$ ikki tub sonning ko‘paytmasi. Shifrlash va deshifrlash — shu halqada darajaga ko‘tarishdir. Masalan, $n = 3233 = 61 \cdot 53$, ochiq daraja $e = 17$ bo‘lsa, $m = 65$ xabar $c = 65^{17} \bmod 3233 = 2790$ ko‘rinishida shifrlanadi. Halqaning tuzilishi va Eyler teoremasi deshifrlash mumkinligini kafolatlaydi.

**AES algoritmi** esa $256$ elementli chekli **maydon** $\mathrm{GF}(2^8)$ ustida quriladi. Bu maydon $\mathbb{Z}_2$ ustidagi ko‘phadlar halqasini keltirilmaydigan $8$-darajali ko‘phad — masalan $\varphi(x) = x^8 + x^4 + x^3 + x + 1$ — bo‘yicha bo‘lgandagi qoldiqlardan iborat.<sup>8</sup> AES ning `SubBytes` bosqichida har bir baytning shu maydondagi **ko‘paytmaviy teskarisi** olinadi — bu amal aynan maydon bo‘lgani (ya’ni har bir noldan farqli baytning teskarisi mavjud bo‘lgani) uchungina ma’noga ega.

Aniq sonli misol: AES maydonida $\mathtt{0x57}$ va $\mathtt{0x83}$ baytlarni ko‘paytirsak, natija $\mathtt{0xC1}$ bo‘ladi. Oddiy butun sonlarda $\mathtt{0x57} \cdot \mathtt{0x83}$ butunlay boshqa (va bir baytga sig‘maydigan) son bo‘lardi; maydon strukturasi esa natijani doimo bitta bayt ($0$–$255$) ichida ushlab turadi — bu yopiqlik xossasining amaliy ko‘rinishidir.


In [4]:
from math import gcd

def maydonmi(p):
    """Z_p maydon bo‘lishini tekshiradi: har bir noldan farqli element teskariga egami?"""
    for a in range(1, p):
        if not any((a * b) % p == 1 for b in range(1, p)):
            return False
    return True

for p in [5, 6, 7, 8, 11]:
    holat = "MAYDON" if maydonmi(p) else "maydon EMAS"
    print(f"Z_{p:<2}: {holat}   (tub son: {all(p % d for d in range(2, p))})")

print("\n--- GF(2^8) da ko‘paytirish (AES maydoni) ---")
def gf_mul(a, b, mod=0x11B):   # 0x11B = x^8 + x^4 + x^3 + x + 1
    natija = 0
    while b:
        if b & 1:
            natija ^= a          # qo‘shish = XOR
        b >>= 1
        a <<= 1
        if a & 0x100:            # daraja 8 ga yetsa, keltirilmaydigan ko‘phadga bo‘lamiz
            a ^= mod
    return natija & 0xFF

x, y = 0x57, 0x83
print(f"0x{x:02X} · 0x{y:02X} = 0x{gf_mul(x, y):02X}  (kutilgan: 0xC1)")

# Har bir noldan farqli bayt teskariga egaligini tekshiramiz (maydon belgisi)
teskarilar = {a: next(b for b in range(1, 256) if gf_mul(a, b) == 1) for a in range(1, 256)}
print(f"Barcha 255 ta noldan farqli baytning teskarisi mavjud: {len(teskarilar) == 255}")
print(f"Masalan, 0x57 ning teskarisi = 0x{teskarilar[0x57]:02X}, "
      f"tekshirish: 0x57 · 0x{teskarilar[0x57]:02X} = 0x{gf_mul(0x57, teskarilar[0x57]):02X}")

Z_5 : MAYDON   (tub son: True)
Z_6 : maydon EMAS   (tub son: False)
Z_7 : MAYDON   (tub son: True)
Z_8 : maydon EMAS   (tub son: False)
Z_11: MAYDON   (tub son: True)

--- GF(2^8) da ko‘paytirish (AES maydoni) ---
0x57 · 0x83 = 0xC1  (kutilgan: 0xC1)
Barcha 255 ta noldan farqli baytning teskarisi mavjud: True
Masalan, 0x57 ning teskarisi = 0xBF, tekshirish: 0x57 · 0xBF = 0x01


## 2.5 Kompleks sonlar maydoni

Haqiqiy sonlar orasida $x^2 = -1$ tenglamaning yechimi yo‘q, chunki har qanday haqiqiy sonning kvadrati manfiy emas. Matematiklar bu kamchilikni bartaraf etish uchun sonlar to‘plamini kengaytirib, kvadrati $-1$ ga teng bo‘lgan yangi **mavhum birlik** $i$ ni kiritdilar:

$$i^2 = -1.$$

**Ta’rif.** **Kompleks sonlar** to‘plami — bu

$$\mathbb{C} = \{a + bi \mid a, b \in \mathbb{R}\}$$

ko‘rinishidagi ifodalar to‘plami, bunda $i^2 = -1$.<sup>6</sup> $z = a + bi$ da $a$ — sonning **haqiqiy qismi** ($\operatorname{Re} z$), $b$ — **mavhum qismi** ($\operatorname{Im} z$) deyiladi.

### Amallar

Kompleks sonlar ustidagi amallar oddiy algebraik qoidalar bo‘yicha, faqat $i^2 = -1$ ni hisobga olgan holda bajariladi.<sup>6</sup>

**Qo‘shish** — haqiqiy va mavhum qismlarni alohida qo‘shish:

$$(a + bi) + (c + di) = (a + c) + (b + d)i.$$

**Ko‘paytirish** — ko‘phadlarni ko‘paytirgandek, so‘ng $i^2$ ni $-1$ ga almashtirish:

$$(a + bi)(c + di) = ac + adi + bci + bdi^2 = (ac - bd) + (ad + bc)i.$$

Har bir $z = a + bi$ uchun ikkita muhim kattalik aniqlanadi.<sup>6</sup> **Kompleks-qo‘shma** son — mavhum qismning ishorasini o‘zgartirish natijasi:

$$\bar{z} = a - bi.$$

**Modul (absolyut qiymat)** — sonning koordinatalar tekisligidagi boshlang‘ichdan uzoqligi:

$$|z| = \sqrt{a^2 + b^2}.$$

Bu ikki kattalik $z \cdot \bar{z} = (a+bi)(a-bi) = a^2 + b^2 = |z|^2$ munosabat orqali bog‘langan. Aynan shu tenglik **teskari element** formulasini beradi: agar $z \ne 0$ bo‘lsa,

$$z^{-1} = \frac{\bar{z}}{|z|^2} = \frac{a - bi}{a^2 + b^2}.$$<sup>6</sup>

**Misol.** $z = 2 + 3i$, $w = 1 - 2i$ bo‘lsin.<sup>6</sup> U holda:

$$z + w = (2+1) + (3-2)i = 3 + i,$$
$$zw = (2 \cdot 1 - 3 \cdot (-2)) + (2 \cdot (-2) + 3 \cdot 1)i = (2 + 6) + (-4 + 3)i = 8 - i,$$
$$|z| = \sqrt{2^2 + 3^2} = \sqrt{13}, \qquad z^{-1} = \frac{2 - 3i}{13} = \frac{2}{13} - \frac{3}{13}i.$$

### Nima uchun $\mathbb{C}$ — maydon?

Kompleks sonlar odatdagi qo‘shish va ko‘paytirishga nisbatan **maydon** tashkil etadi.<sup>4, 8</sup> Buni ma’ruzada quyidagicha tekshirilgan:<sup>8</sup>

1. **Birlik element.** $\mathbb{C}$ da ko‘paytirishning birlik elementi $1 = 1 + 0i$; qo‘shishning noli esa $0 = 0 + 0i$.
2. **Teskari element.** Har bir $z = a + bi \ne 0$ uchun yuqoridagi $z^{-1} = \dfrac{a - bi}{a^2 + b^2}$ mavjud — chunki $z \ne 0$ bo‘lsa $a^2 + b^2 > 0$, ya’ni maxrajda hech qachon nol turmaydi.
3. **Nolning bo‘luvchilari yo‘q.** Agar $z \cdot w = 0$ bo‘lsa, modullar orqali $|z| \cdot |w| = 0$; haqiqiy sonlar orasida ko‘paytma nol bo‘lishi uchun ko‘paytuvchilardan biri nol bo‘lishi kerak, demak $z = 0$ yoki $w = 0$.

Uch shartning bajarilishi — birlikning mavjudligi, teskari elementning mavjudligi va nol bo‘luvchilarning yo‘qligi — $\mathbb{C}$ ning maydon ekanligini ko‘rsatadi.

### Nega bu axborot xavfsizligida muhim?

Kompleks sonlar maydonining o‘zi to‘g‘ridan-to‘g‘ri shifrlashda kalit sifatida ishlatilmaydi (kriptografiya, aksincha, **chekli** maydonlarni afzal ko‘radi). Ammo kompleks sonlar ustidagi **birlik ildizlari** g‘oyasi zamonaviy kriptografiyada juda muhim rol o‘ynaydi.

Panjaraga asoslangan post-kvant algoritmlarda (masalan, standartlashtirilgan **Kyber** va **Dilithium** sxemalarida) ko‘phadlarni tez ko‘paytirish uchun **Sonli-nazariy almashtirish (NTT)** qo‘llaniladi. NTT — bu klassik **Diskret Fure almashtirishi (DFT)** ning chekli maydonga ko‘chirilgan ko‘rinishi; DFT esa aynan **birlik ildizlari** ($z^n = 1$ tenglama yechimlari) ustiga quriladi. Shunday qilib, quyidagi bo‘limlarda o‘rganadigan birlik ildizlari tushunchasi — minglab ko‘phad koeffitsientlarini samarali ko‘paytirish, natijada esa kalitlarni tez generatsiya qilish imkonini beruvchi mexanizmning matematik yuragidir.


In [5]:
# Kompleks sonlar: Python'da tayyor `complex` tipi (j — mavhum birlik)
z = complex(2, 3)   # 2 + 3i
w = complex(1, -2)  # 1 - 2i
print(f"z = {z},  w = {w}")
print(f"z + w = {z + w}")
print(f"z · w = {z * w}")
print(f"z ning qo‘shmasi (z̄) = {z.conjugate()}")
print(f"|z| = √13 ≈ {abs(z):.6f}")
print(f"z⁻¹ = {1/z}  ≈  2/13 - (3/13)i = {2/13:.6f} - {3/13:.6f}i")

# i² = -1 ni va z·z̄ = |z|² ni tekshiramiz
print("\ni² =", complex(0,1)**2)
print("z · z̄ =", (z * z.conjugate()).real, " = |z|² =", abs(z)**2)

z = (2+3j),  w = (1-2j)
z + w = (3+1j)
z · w = (8-1j)
z ning qo‘shmasi (z̄) = (2-3j)
|z| = √13 ≈ 3.605551
z⁻¹ = (0.15384615384615385-0.23076923076923078j)  ≈  2/13 - (3/13)i = 0.153846 - 0.230769i

i² = (-1+0j)
z · z̄ = 13.0  = |z|² = 12.999999999999998


## 2.6 Kompleks sonning trigonometrik ko‘rinishi

Kompleks sonni $z = a + bi$ ko‘rinishida yozish **algebraik (dekart) ko‘rinish** deyiladi. Bunda son tekislikda gorizontal (haqiqiy) o‘qda $a$, vertikal (mavhum) o‘qda $b$ koordinatali nuqta sifatida tasvirlanadi.<sup>7</sup> Ammo nuqtaning o‘rnini boshqacha — **boshlang‘ichdan uzoqligi** va **burchagi** orqali ham berish mumkin. Bu **qutb (trigonometrik) ko‘rinish**ga olib keladi.

Noldan farqli $z$ nuqtani belgilash uchun ikki kattalik yetarli:<sup>7</sup>

- $r = |z| = \sqrt{a^2 + b^2}$ — **modul**, nuqtaning boshlang‘ichdan uzoqligi;
- $\theta$ — **argument**, musbat haqiqiy o‘qdan nuqtagacha soat mili yo‘nalishiga teskari o‘lchangan burchak.

Oddiy trigonometriyadan $a = r\cos\theta$ va $b = r\sin\theta$. Buni $z = a + bi$ ga qo‘ysak, **trigonometrik ko‘rinish** hosil bo‘ladi:

$$z = r(\cos\theta + i\sin\theta).$$<sup>7</sup>

Bu ifoda ko‘pincha qisqacha $z = r\operatorname{cis}\theta$ deb yoziladi (bunda $\operatorname{cis}\theta$ — “$\cos\theta + i\sin\theta$” ning qisqartmasi). Yozuv bir qiymatli bo‘lishi uchun odatda $0 \le \theta < 2\pi$ (yoki $0^\circ \le \theta < 360^\circ$) deb olinadi.<sup>7</sup>

### Ikki ko‘rinish orasida o‘tish

**Trigonometrikdan algebraikka.** $z = 2\operatorname{cis} 60^\circ$ berilgan bo‘lsin.<sup>7</sup> U holda

$$a = 2\cos 60^\circ = 2 \cdot \tfrac12 = 1, \qquad b = 2\sin 60^\circ = 2 \cdot \tfrac{\sqrt3}{2} = \sqrt3,$$

demak $z = 1 + \sqrt3\,i$.

**Algebraikdan trigonometrikka.** $z = 3\sqrt2 - 3\sqrt2\,i$ berilgan bo‘lsin.<sup>7</sup> U holda

$$r = \sqrt{(3\sqrt2)^2 + (3\sqrt2)^2} = \sqrt{18 + 18} = \sqrt{36} = 6,$$
$$\theta = \arctan\!\left(\frac{-3\sqrt2}{3\sqrt2}\right) = \arctan(-1) = 315^\circ,$$

demak $z = 6\operatorname{cis} 315^\circ$. Burchakni topishda nuqtaning qaysi chorakda ekanini hisobga olish muhim: bu yerda $a > 0$, $b < 0$ — nuqta to‘rtinchi chorakda, shuning uchun $315^\circ$ ni tanlaymiz.

### Nima uchun trigonometrik ko‘rinish qulay?

Trigonometrik ko‘rinishning asosiy afzalligi — **ko‘paytirish va darajaga ko‘tarishni nihoyatda soddalashtirishi**.<sup>7</sup> Ikki kompleks sonni ko‘paytirganda modullar ko‘paytiriladi, argumentlar esa qo‘shiladi:

$$z = r\operatorname{cis}\theta, \quad w = s\operatorname{cis}\varphi \quad\Longrightarrow\quad zw = rs\operatorname{cis}(\theta + \varphi).$$

**Misol.** $z = 3\operatorname{cis}\tfrac{\pi}{3}$ va $w = 2\operatorname{cis}\tfrac{\pi}{6}$ bo‘lsa,<sup>7</sup>

$$zw = 3 \cdot 2 \operatorname{cis}\!\left(\tfrac{\pi}{3} + \tfrac{\pi}{6}\right) = 6\operatorname{cis}\tfrac{\pi}{2} = 6i.$$

Geometrik ma’nosi juda tabiiy: kompleks songa ko‘paytirish — tekislikni **cho‘zish (modul bo‘yicha) va burish (argument bo‘yicha)** demakdir. Aynan shu geometrik talqin keyingi bo‘limdagi daraja va ildiz formulalarini oson tushunishga yordam beradi.

## 2.7 Kompleks sonni darajaga ko‘tarish va ildizdan chiqarish

### De Muavr formulasi (darajaga ko‘tarish)

Ko‘paytirishda argumentlar qo‘shilishini bir sonni o‘z-o‘ziga takror ko‘paytirishga qo‘llasak, darajaga ko‘tarishning ajoyib sodda formulasi kelib chiqadi.

**Teorema (De Muavr).** Ixtiyoriy noldan farqli $z = r\operatorname{cis}\theta$ kompleks son va har qanday natural $n$ uchun

$$z^n = \big[r(\cos\theta + i\sin\theta)\big]^n = r^n\big(\cos n\theta + i\sin n\theta\big) = r^n\operatorname{cis}(n\theta).$$<sup>7</sup>

Ya’ni: **modul $n$-darajaga ko‘tariladi, argument esa $n$ ga ko‘paytiriladi**. Buni matematik induksiya bilan isbotlash mumkin: $n = 1$ da aniq to‘g‘ri; agar $z^n = r^n\operatorname{cis}(n\theta)$ o‘rinli bo‘lsa, $z^{n+1} = z^n \cdot z = r^n\operatorname{cis}(n\theta) \cdot r\operatorname{cis}\theta = r^{n+1}\operatorname{cis}((n+1)\theta)$ — bu yerda ko‘paytirish qoidasi (modullar ko‘payadi, argumentlar qo‘shiladi) ishlatildi.<sup>7</sup>

**Misol.** $z = 1 + i$ uchun $z^{10}$ ni hisoblaymiz.<sup>7</sup> To‘g‘ridan-to‘g‘ri $(1+i)^{10}$ ni yoyish mashaqqatli. Avval qutb ko‘rinishiga o‘tamiz: $r = \sqrt{1^2 + 1^2} = \sqrt2$, $\theta = 45^\circ = \tfrac{\pi}{4}$, ya’ni $z = \sqrt2\operatorname{cis}\tfrac{\pi}{4}$. Endi De Muavr formulasi bo‘yicha:

$$z^{10} = (\sqrt2)^{10}\operatorname{cis}\!\left(10 \cdot \tfrac{\pi}{4}\right) = 32\operatorname{cis}\tfrac{5\pi}{2} = 32\operatorname{cis}\tfrac{\pi}{2} = 32i.$$

(Bu yerda $\tfrac{5\pi}{2}$ dan to‘liq aylanish $2\pi$ ni chiqarib, $\tfrac{\pi}{2}$ ga keltirdik.)

### Ildizdan chiqarish va birlik ildizlari

Darajaga ko‘tarish argumentni **ko‘paytirsa**, ildizdan chiqarish uni **bo‘lishi** kutiladi. Ammo bu yerda nozik nuqta bor: argument $2\pi$ ga karrali qismgacha aniqlangani uchun, $n$-darajali ildiz **bitta emas, balki $n$ ta** turli qiymatga ega bo‘ladi.

Eng muhim maxsus hol — **birlikning $n$-darajali ildizlari**, ya’ni $z^n = 1$ tenglamaning yechimlari.

**Teorema.** $z^n = 1$ tenglamaning barcha yechimlari (birlikning $n$-darajali ildizlari)

$$z_k = \operatorname{cis}\!\left(\frac{2\pi k}{n}\right) = \cos\frac{2\pi k}{n} + i\sin\frac{2\pi k}{n}, \qquad k = 0, 1, \dots, n-1$$

ko‘rinishida bo‘ladi.<sup>7</sup> Bu $n$ ta ildiz birlik radiusli aylanada **teng oraliqda** joylashadi va ko‘paytirishga nisbatan $n$-tartibli **siklik gruppa** tuzadi.<sup>7</sup>

De Muavr yordamida tekshirish oson: $z_k^n = \operatorname{cis}\!\left(n \cdot \tfrac{2\pi k}{n}\right) = \operatorname{cis}(2\pi k) = 1$, chunki har qanday $2\pi k$ burchak to‘liq aylanishlardan iborat.

**Misol.** $n = 4$ uchun birlikning to‘rtta ildizi:

$$z_0 = \operatorname{cis} 0 = 1, \quad z_1 = \operatorname{cis}\tfrac{\pi}{2} = i, \quad z_2 = \operatorname{cis}\pi = -1, \quad z_3 = \operatorname{cis}\tfrac{3\pi}{2} = -i.$$

Ular $\{1, i, -1, -i\}$ — 2.3-bo‘limda uchragan $\mathbb{C}^{*}$ ning kichik gruppasi bilan aynan bir xil.<sup>2</sup> Bu tasodif emas: birlik ildizlari algebraik struktura (gruppa) va geometriya (aylanadagi teng nuqtalar) o‘rtasidagi ko‘prikdir.

### Nega bu axborot xavfsizligida muhim?

Yuqorida aytilganidek, birlik ildizlari **Diskret Fure almashtirishi (DFT)** va uning tez varianti **FFT** ning asosidir; ularning chekli maydondagi analogi — **NTT** — post-kvant kriptografiyada ko‘phadlarni tez ko‘paytirish uchun ishlatiladi.

Buning ahamiyatini sonlar bilan tasavvur qilaylik: ikkita $n = 256$ darajali ko‘phadni oddiy usulda ko‘paytirish taxminan $n^2 = 256^2 = 65\,536$ ta ko‘paytirish talab qiladi. NTT/FFT esa buni taxminan $n\log_2 n = 256 \cdot 8 = 2\,048$ ta amalga tushiradi — ya’ni deyarli **30 barobar** tezroq. Kyber kabi sxemalarda ana shunday ko‘phad ko‘paytirishlari kalit generatsiyasi va shifrlashda minglab marta bajarilishini hisobga olsak, birlik ildizlariga asoslangan bu tezlashtirish butun tizimni amaliy jihatdan qo‘llansa bo‘ladigan darajaga chiqaradi.


In [6]:
import cmath, math

# Algebraik ↔ trigonometrik o‘tish
z = complex(3*math.sqrt(2), -3*math.sqrt(2))
r, theta = abs(z), cmath.phase(z) % (2*math.pi)   # modul va argument, [0,2π) oraliqda
print(f"z = {z.real:.3f} + ({z.imag:.3f})i")
print(f"modul r = {r:.3f},  argument θ = {math.degrees(theta):.1f}°  → z = {r:.0f}·cis({math.degrees(theta):.0f}°)")

# De Muavr: (1+i)^10
z = complex(1, 1)
r, theta = abs(z), cmath.phase(z)
z10_demuavr = r**10 * cmath.exp(1j*10*theta)   # r^10 · cis(10θ)
print(f"\n(1+i)^10 = {(z**10).real:.0f} + ({(z**10).imag:.0f})i   [to‘g‘ridan-to‘g‘ri]")
print(f"        = {z10_demuavr.real:.0f} + ({z10_demuavr.imag:.0f})i   [De Muavr: 32·cis(π/2)=32i]")

# Birlikning n-darajali ildizlari
n = 4
print(f"\nBirlikning {n} ta ildizi (z^{n} = 1):")
for k in range(n):
    zk = cmath.exp(2j*math.pi*k/n)
    print(f"  z_{k} = cis(2π·{k}/{n}) = {zk.real:+.0f} {zk.imag:+.0f}i   →  z_{k}^{n} = {round((zk**n).real):+.0f}")

z = 4.243 + (-4.243)i
modul r = 6.000,  argument θ = 315.0°  → z = 6·cis(315°)

(1+i)^10 = 0 + (32)i   [to‘g‘ridan-to‘g‘ri]
        = 0 + (32)i   [De Muavr: 32·cis(π/2)=32i]

Birlikning 4 ta ildizi (z^4 = 1):
  z_0 = cis(2π·0/4) = +1 +0i   →  z_0^4 = +1
  z_1 = cis(2π·1/4) = +0 +1i   →  z_1^4 = +1
  z_2 = cis(2π·2/4) = -1 +0i   →  z_2^4 = +1
  z_3 = cis(2π·3/4) = -0 -1i   →  z_3^4 = +1


## Mashqlar

1. $\mathbb{Z}_6 = \{0, 1, 2, 3, 4, 5\}$ to‘plamida mod $6$ qo‘shish amalining amallar jadvalini tuzing. Bu struktura gruppami? $4$ elementning teskarisini toping.
2. $\mathbb{N} = \{1, 2, 3, \dots\}$ ko‘paytirishga nisbatan: (a) yarim gruppami? (b) monoidmi? (c) gruppami? Har bir javobingizni asoslang.
3. $U(10)$ ni (ya’ni $10$ bilan o‘zaro tub qoldiqlarni) yozing. Bu to‘plam mod $10$ ko‘paytirishga nisbatan gruppa ekanini tekshiring va $3$ ning teskarisini toping.
4. Nima uchun $\mathbb{Z}_9$ maydon emas? Nol bo‘luvchilarga bitta aniq misol keltiring.
5. Diffie–Hellman sxemasida $p = 23$, $g = 5$, Alisaning maxfiy soni $a = 4$, Bobning maxfiy soni $b = 3$ bo‘lsin. Ochiq uzatiladigan $A$, $B$ larni va umumiy maxfiy kalitni hisoblang.
6. $z = 3 + 4i$ va $w = 1 - 2i$ uchun quyidagilarni toping: $z + w$, $zw$, $\bar{z}$, $|z|$, $z^{-1}$.
7. $z = -1 + \sqrt3\,i$ ni trigonometrik ko‘rinishga o‘tkazing (modul va argumentni toping).
8. De Muavr formulasidan foydalanib $(1 + i)^{8}$ ni hisoblang.
9. Birlikning $6$-darajali barcha ildizlarini ($z^6 = 1$) trigonometrik ko‘rinishda yozing va ulardan uchtasini $a + bi$ ko‘rinishga o‘tkazing.

## Javoblar

1. Jadval $\mathbb{Z}_4$ dagiga o‘xshash, faqat $6 \times 6$. Bu — Abel gruppasi (mod $6$ qo‘shish barcha to‘rt shartni qanoatlantiradi). $4$ ning teskarisi $2$, chunki $4 + 2 = 6 \equiv 0 \pmod 6$.
2. (a) Ha, yarim gruppa — ko‘paytirish yopiq va assotsiativ. (b) Ha, monoid — neytral element $1$ mavjud. (c) Yo‘q, gruppa emas — masalan $2$ ning ko‘paytmaviy teskarisi $\tfrac12$ natural son emas.
3. $U(10) = \{1, 3, 7, 9\}$. Bularning har biri $10$ bilan o‘zaro tub; jadval yopiq va har bir element teskariga ega, demak gruppa. $3$ ning teskarisi $7$, chunki $3 \cdot 7 = 21 \equiv 1 \pmod{10}$.
4. $9 = 3 \cdot 3$ — tub emas. $\mathbb{Z}_9$ da $3 \cdot 3 = 9 \equiv 0$, ya’ni noldan farqli ikki elementning ko‘paytmasi nol. Nol bo‘luvchilari borligi uchun $\mathbb{Z}_9$ butunlik sohasi ham, maydon ham emas.
5. $A = 5^4 \bmod 23 = 4$; $B = 5^3 \bmod 23 = 10$. Umumiy kalit: $B^a = 10^4 \bmod 23 = 18$ va $A^b = 4^3 \bmod 23 = 18$ — mos keladi, kalit $18$.
6. $z + w = 4 + 2i$; $zw = (3 \cdot 1 - 4 \cdot (-2)) + (3 \cdot (-2) + 4 \cdot 1)i = 11 - 2i$; $\bar{z} = 3 - 4i$; $|z| = \sqrt{9 + 16} = 5$; $z^{-1} = \dfrac{3 - 4i}{25} = \dfrac{3}{25} - \dfrac{4}{25}i$.
7. $r = \sqrt{(-1)^2 + (\sqrt3)^2} = \sqrt{4} = 2$. Nuqta ikkinchi chorakda ($a < 0$, $b > 0$), argument $\theta = 120^\circ = \tfrac{2\pi}{3}$. Demak $z = 2\operatorname{cis} 120^\circ$.
8. $1 + i = \sqrt2\operatorname{cis} 45^\circ$, shuning uchun $(1+i)^8 = (\sqrt2)^8\operatorname{cis}(8 \cdot 45^\circ) = 16\operatorname{cis} 360^\circ = 16\operatorname{cis} 0^\circ = 16$.
9. $z_k = \operatorname{cis}\dfrac{2\pi k}{6} = \operatorname{cis}(60^\circ k)$, $k = 0, 1, \dots, 5$. Ular: $z_0 = 1$, $z_1 = \tfrac12 + \tfrac{\sqrt3}{2}i$, $z_2 = -\tfrac12 + \tfrac{\sqrt3}{2}i$, $z_3 = -1$, $z_4 = -\tfrac12 - \tfrac{\sqrt3}{2}i$, $z_5 = \tfrac12 - \tfrac{\sqrt3}{2}i$.

## Snoskalar (manbalar)

<sup>1</sup> Shoup V. “A Computational Introduction to Number Theory and Algebra” (2-nashr), “Preliminaries” bo‘limi, “Binary operations” — binar amal ta’rifi, yopiqlik (closed under), assotsiativlik va kommutativlik ta’riflari.

<sup>2</sup> Judson T. W. “Abstract Algebra: Theory and Applications”, 3-bob “Groups”, §3.2 “Definitions and Examples” — gruppa ta’rifi (uch aksioma), Abel (kommutativ) gruppa, $(\mathbb{Z}, +)$, $(\mathbb{Z}_n, +)$, $U(n)$ misollari, Keli jadvali; §3.3 “Subgroups” — $\{1, -1, i, -i\}$ va $\mathrm{SL}_2$ misollari.

<sup>3</sup> Shoup V. “A Computational Introduction to Number Theory and Algebra” (2-nashr), 6-bob “Abelian groups”, §6.1 — Ta’rif 6.1 (Abel gruppa aksiomalari), Teorema 6.2 (neytral va teskari elementning yagonaligi), Misol 6.10–6.14 (bitlar XOR gruppasi, satrlarni birlashtirish monoidi, Abel bo‘lmagan gruppalar).

<sup>4</sup> Judson T. W. “Abstract Algebra: Theory and Applications”, 16-bob “Rings”, §16.1–16.2 — halqa ta’rifi (olti aksioma), birlikli halqa, kommutativ halqa, butunlik sohasi, bo‘lish halqasi va maydon ta’riflari, $\mathbb{Z}$, $\mathbb{Q}$, $\mathbb{Z}_p$, Gauss butun sonlari misollari.

<sup>5</sup> Shoup V. “A Computational Introduction to Number Theory and Algebra” (2-nashr), 7-bob “Rings”, Ta’rif 7.1 — birlikli kommutativ halqa ta’rifi (qo‘shishga nisbatan Abel gruppa, ko‘paytirishning assotsiativligi, distributivlik, birlik, kommutativlik).

<sup>6</sup> Judson T. W. “Abstract Algebra: Theory and Applications”, 4-bob “Cyclic Groups”, §4.2 “Multiplicative Group of Complex Numbers” — $\mathbb{C}$ ta’rifi, qo‘shish va ko‘paytirish, kompleks-qo‘shma, modul, teskari element formulasi, $z = 2 + 3i$, $w = 1 - 2i$ misoli.

<sup>7</sup> Judson T. W. “Abstract Algebra: Theory and Applications”, 4-bob, §4.2 — qutb (trigonometrik) ko‘rinish $r\operatorname{cis}\theta$, ko‘paytirish qoidasi (Taklif 4.20), De Muavr teoremasi (Teorema 4.22) va $(1+i)^{10}$ misoli, birlik ildizlari (Teorema 4.25).

<sup>8</sup> “Matematik strukturalar. Gruppa, halqa, maydon va uning xossalari” hamda “AES algoritmining matematik asoslari” ma’ruza matnlari (25.10.2023–22.11.2023) — strukturalarning zinapoyasimon ketma-ketligi (yarim gruppa → monoid → gruppa), $(\mathbb{Z}, +, \cdot)$ va $(\mathbb{C}, +, \cdot)$ ni tekshirish, maydon xossalari, $\mathrm{GF}(2^8)$ va AES `SubBytes` bilan bog‘lanish.
